### PR TITLE
iox-#1196 Remove conversion operator in `RelativePointer`

### PR DESCRIPTION
**Pull request by mossmaurice**
## Pre-Review Checklist for the PR Author

1. [x] Code follows the coding style of [CONTRIBUTING.md][contributing]
1. [x] Tests follow the [best practice for testing][testing]
1. [x] ~~Changelog updated [in the unreleased section][changelog] including API breaking changes~~
1. [x] Branch follows the naming format (`iox-#123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit messages are signed (`git commit -s`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed (except `task-list-completed`)
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[contributing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#coding-style
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/concepts/best-practice-for-testing.md
[changelog]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/release-notes/iceoryx-unreleased.md

## Notes for Reviewer

* Instead of making the conversion operator `explicit` remove it and use `get()`

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Code according to our coding style and naming conventions
- [x] Unit tests have been written for new behavior
    - [x] Each unit test case has a unique UUID
- [x] Public API changes are documented via doxygen
- [x] Copyright owner are updated in the changed files
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [x] All open points are addressed and tracked via issues

## References

- Closes #1196 (partly)

### CODE REVIEW

File: iceoryx_hoofs/source/concurrent/loffli.cpp [hunk 1]
```diff
@@ -42,7 +42,7 @@ void LoFFLi::init(cxx::not_null<Index_t*> freeIndicesMemory, const uint32_t capa
         for (uint32_t i = 0; i < m_size + 1; i++)
         {
             // NOLINTNEXTLINE(cppcoreguidelines-pro-bounds-pointer-arithmetic) upper limit of index is set by m_size
-            m_nextFreeIndex[i] = i + 1;
+            m_nextFreeIndex.get()[i] = i + 1;
```
Comment (MatthiasKillat) [inline iceoryx_hoofs/source/concurrent/loffli.cpp:45]: I will review later, but just looking at this syntax it already feels wrong to remove the operator. It is not really clear from reading either, just more verbose and confusing.

I clearly remember why I opted for the implicit conversion to pointer. If I read what is originally there I just think: "sure, it is some kind of pointer and we use pointer arithmetic and dereference it." For the understanding at this level it does not matter what kind of pointer it has, only that it has pointer syntax (and the corresponding access semantics).

But now I pause a little and need to think about that I probably get some pointer back and then we perform the pointer arithmetic. It would even look worse with explicit conversion.

For me it is the posterchild case where implicit conversion is the way to go.

Comment (MatthiasKillat) [reply iceoryx_hoofs/source/concurrent/loffli.cpp:45]: Sure, making it explicit is worse and I do not advocate this. I advocate keeping it implicit as it shall behave like a pointer from the point of view of the client,  which is exactly what is accomplished by implicit conversion.

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/relocatable_pointer/relative_pointer.hpp [hunk 2]
```diff
@@ -82,9 +82,19 @@ class RelativePointer : public BaseRelativePointer
...
-    /// @return a pointer of type T pointing to the underlying object
-    operator T*() const noexcept;
+    /// @brief converts the RelativePointer to bool
+    /// @return bool which contains true if the RelativePointer contains a pointer
+    explicit operator bool() const noexcept;
+
+    /// @brief True if the underyling pointer is smaller than the one in the rhs
+    /// @param[in] rhs Another RelativePointer
+    /// @return True if smaller, otherwise false
+    bool operator<(const RelativePointer<T>& rhs) const noexcept;
```
Comment (MatthiasKillat) [inline iceoryx_hoofs/include/iceoryx_hoofs/internal/relocatable_pointer/relative_pointer.hpp:92]: Do we need these operators? Usually pointer comparison like this is a bad idea. Equality comparison is useful, but order not so much. One reason is that a order has to satisfy certain properties (basically be at least a irreflexive transitive relation (this is implicitly assumed by various STL constructs).

E.g. `std::map<K,V>` requires that these operators exist for K, but I am not aware that we use these pointers in a map. If the relation < exists but does not have these properties, weird things happen.

One can define it to essentially be the < relation on the corresponding raw pointer.
However, it makes little sense to compare these kinds of pointers as the relation cannot hold across process boundaries (i.e. in one process the pointers may compare as <, in the other as >). This may lead to all kinds of unpredictable problems.

Implicitly this was the case before as well, but not really responsibility of the class.

TLDR: Expecting an order on raw pointers is a bad idea. Expecting an order on fancy pointers is worse (especially those that have to work across process boundaries). The operators should be removed unless really needed. If the user wants to compare the raw pointers after a `get` she should do just that (and be aware of the problems).

**Review by MatthiasKillat — COMMENTED** [review]
implicit vs. explicit conversion aside (I prefer implicit conversion here), there is the issue whether `operator<` should exist here. If we do not compare those pointers in such a way, it should be removed. If we do, the question is: why? (it was previously possible due to implicit conversion and compared the raw pointer). pointer comparison with < is usually not a good idea.

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/relocatable_pointer/base_relative_pointer.hpp [hunk 3]
```diff
@@ -174,6 +174,8 @@ class BaseRelativePointer
     static constexpr offset_t NULL_POINTER_OFFSET = std::numeric_limits<offset_t>::max();
 
   protected:
+    ~BaseRelativePointer() noexcept = default;
```
Comment (MatthiasKillat) [inline iceoryx_hoofs/include/iceoryx_hoofs/internal/relocatable_pointer/base_relative_pointer.hpp:177]: Update copyright.

File: iceoryx_hoofs/source/concurrent/loffli.cpp [hunk 4]
```diff
@@ -42,7 +42,7 @@ void LoFFLi::init(cxx::not_null<Index_t*> freeIndicesMemory, const uint32_t capa
         for (uint32_t i = 0; i < m_size + 1; i++)
```
Comment (MatthiasKillat) [inline iceoryx_hoofs/source/concurrent/loffli.cpp:42]: Update copyright.

File: iceoryx_posh/source/mepoo/mem_pool.cpp [hunk 5]
```diff
@@ -90,15 +90,15 @@ void* MemPool::getChunk() noexcept
     m_usedChunks.fetch_add(1U, std::memory_order_relaxed);
     adjustMinFree();
```
Comment (MatthiasKillat) [inline iceoryx_posh/source/mepoo/mem_pool.cpp:92]: Update copyright.

File: iceoryx_posh/source/mepoo/shm_safe_unmanaged_chunk.cpp [hunk 6]
```diff
@@ -59,7 +59,7 @@ SharedChunk ShmSafeUnmanagedChunk::releaseToSharedChunk() noexcept
     }
     auto chunkMgmt = rp::RelativePointer<mepoo::ChunkManagement>(m_chunkManagement.offset(), m_chunkManagement.id());
     m_chunkManagement.reset();
```
Comment (MatthiasKillat) [inline iceoryx_posh/source/mepoo/shm_safe_unmanaged_chunk.cpp:61]: Update copyright.

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/relocatable_pointer/relative_pointer.hpp (hunk 2 above)
Comment (mossmaurice) [reply iceoryx_hoofs/include/iceoryx_hoofs/internal/relocatable_pointer/relative_pointer.hpp:92]: We store `RelativePointer`s in `cxx::vector`. The operators are required for `std::sort` in file `iceoryx_posh/internal/popo/building_blocks/chunk_distributor.inl:179-180`. The code was added for the blocking publisher feature. I understand your point and I'm not sure if we uncovered a conversion problem here. My understanding:

```
// chunk_distributor_data.hpp
using QueueContainer_t =
    cxx::vector<rp::RelativePointer<ChunkQueueData_t>, ChunkDistributorDataProperties_t::MAX_QUEUES>;

// chunk_distributor.inl
std::sort(getMembers()->m_queues.begin(), getMembers()->m_queues.end());

// vector.inl
using const_iterator = const T*; // with T being RelativePointer
return reinterpret_cast<const_iterator>(&at_unchecked(0));

This used to call the conversion operator and use the `operator<(const ChunkQueueData& lhs, ChunkQueueData rhs)`. This is now done via the new operators explicitly.
```

Comment (MatthiasKillat) [reply iceoryx_hoofs/include/iceoryx_hoofs/internal/relocatable_pointer/relative_pointer.hpp:92]: The question is what should be sorted here? If its is just to have a well-defined order it is fine to sort by pointer. In fact I think sorting by data here is not well-defined anyway.

The elements point to something in shared memory, which in a fixed process address space can act as unique identifiers and are sortable by  pointer. I just think that the solution is to use an explicit sorting criterion with `std::sort`, e.g. with a functor struct or function that defines the order.

Comment (mossmaurice) [reply iceoryx_hoofs/include/iceoryx_hoofs/internal/relocatable_pointer/relative_pointer.hpp:92]: @elfenpiff Can you answer the above question? Looking again at the code, I think we want to sort the pointers, correct?

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/relocatable_pointer/base_relative_pointer.hpp [hunk 7]
```diff
@@ -1,5 +1,5 @@
 // Copyright (c) 2019 by Robert Bosch GmbH. All rights reserved.
-// Copyright (c) 2021 by Apex.AI Inc. All rights reserved.
```
Comment (elfenpiff) [inline iceoryx_hoofs/include/iceoryx_hoofs/internal/relocatable_pointer/base_relative_pointer.hpp:2]: Please add all touched files from hoofs into `.clang-tidy-diff-scans.txt`

**Review by elfenpiff — CHANGES_REQUESTED** [review]

File: iceoryx_hoofs/include/iceoryx_hoofs/internal/relocatable_pointer/relative_pointer.hpp (hunk 2 above)
Comment (mossmaurice) [reply iceoryx_hoofs/include/iceoryx_hoofs/internal/relocatable_pointer/relative_pointer.hpp]: Talked to @elfenpiff the other day, the sorting mechanism is indeed wanted here and correct.

File: iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_distributor.inl [hunk 8]
```diff
@@ -176,14 +176,19 @@ inline uint64_t ChunkDistributor<ChunkDistributorDataType>::deliverToAllStoredQu
...
-            std::sort(getMembers()->m_queues.begin(), getMembers()->m_queues.end());
-            std::sort(remainingQueues.begin(), remainingQueues.end());
+            auto greaterThan = [](rp::RelativePointer<ChunkQueueData_t>& a,
+                                  rp::RelativePointer<ChunkQueueData_t>& b) -> bool {
+                return reinterpret_cast<uint64_t>(a.get()) > reinterpret_cast<uint64_t>(b.get());
+            };
+            std::sort(getMembers()->m_queues.begin(), getMembers()->m_queues.end(), greaterThan);
+            std::sort(remainingQueues.begin(), remainingQueues.end(), greaterThan);
 
             auto iter = std::set_intersection(getMembers()->m_queues.begin(),
```
Comment (MatthiasKillat) [inline iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_distributor.inl:186]: I double checked, and it appears that `set_intersection` requires sorted ranges despite in theory being able to sort them itself (by the predicate provided). It is important that the ranges are sorted by the same predicate used in the intersection, which is the case here.

For us it does not even matter much what the order is, just that it is a well-defined order (irreflexive, transitive etc.), which it is if sorting.

**Review by MatthiasKillat — DISMISSED** [review]
Functionally it looks fine, but I still think that implicit conversion and suppression of static analysis findings would be the way to go here (reasons where stated in comments).

File: iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_queue_popper.inl [hunk 9]
```diff
@@ -145,7 +145,7 @@ inline void ChunkQueuePopper<ChunkQueueDataType>::unsetConditionVariable() noexc
 template <typename ChunkQueueDataType>
 inline bool ChunkQueuePopper<ChunkQueueDataType>::isConditionVariableSet() const noexcept
 {
-    return getMembers()->m_conditionVariableDataPtr;
```
Comment (elBoberido) [inline iceoryx_posh/include/iceoryx_posh/internal/popo/building_blocks/chunk_queue_popper.inl:148]: Why not just `getMembers()->m_conditionVariableDataPtr != nullptr`?

**Review by MatthiasKillat — APPROVED** [review]

**Review by elBoberido — APPROVED** [review]
